File: ios/FiHavenCore/Sources/FiHavenCore/Logic/SubscriptionIcons.swift
import Foundation

/// Map a subscription / merchant name to a recognizable emoji. The web app
/// also renders real SVG brand logos, but native mirrors only the emoji layer
/// (per-brand → category → generic). Keep the brand table in sync with the web
/// `subscriptionIcons.js` and Android `SubscriptionIcons.kt`.
public enum SubscriptionIcons {
    /// Recognizable per-brand emoji, keyed by the normalized name.
    static let brandEmoji: [String: String] = [
        "netflix": "🎬", "disney": "🏰", "disneyplus": "🏰", "hulu": "📺", "max": "📺",
        "hbomax": "📺", "hbo": "📺", "peacock": "🦚", "paramount": "⛰️", "paramountplus": "⛰️",
        "appletv": "🍏", "appletvplus": "🍏", "apple": "🍏",
        "icloud": "☁️", "icloudstorage": "☁️", "icloudplus": "☁️", "googleone": "☁️",
        "dropbox": "📦", "onedrive": "☁️",
        "spotify": "🎵", "applemusic": "🎵", "youtubemusic": "🎵", "pandora": "🎵",
        "tidal": "🎵", "music": "🎵", "deezer": "🎵",
        "youtube": "▶️", "youtubepremium": "▶️", "youtubetv": "📺", "twitch": "🎮",
        "xboxgamepass": "🎮", "playstationplus": "🎮", "playstation": "🎮", "nintendo": "🎮",
        "nintendoswitchonline": "🎮",
        "amazon": "📦", "amazonprime": "📦", "primevideo": "📦",
        "audible": "🎧", "kindle": "📚", "kindleunlimited": "📚",
        "nyt": "📰", "nytimes": "📰", "wsj": "📰", "washingtonpost": "📰", "medium": "📰",
        "theathletic": "📰",
        "patreon": "🅿️", "substack": "📩", "notion": "📝", "evernote": "📝", "obsidian": "📝",
        "adobe": "🎨", "adobecc": "🎨", "canva": "🎨", "figma": "🎨", "lightroom": "🎨",
        "onepassword": "🔐", "bitwarden": "🔐", "lastpass": "🔐", "dashlane": "🔐",
        "nordvpn": "🛡️", "expressvpn": "🛡️", "protonvpn": "🛡️", "proton": "🛡️",
        "chatgpt": "🤖", "openai": "🤖", "claude": "🤖", "anthropic": "🤖", "midjourney": "🤖",
        "perplexity": "🤖",
        "github": "🐙", "githubcopilot": "🐙", "githubpro": "🐙",
        "peloton": "🚲", "strava": "🏃", "calm": "🧘", "headspace": "🧘", "fitbit": "⌚",
        "whoop": "⌚",
        "crunchyroll": "🍥", "openbubbles": "💬", "slack": "💬", "discord": "💬", "zoom": "🎥",
        "microsoft365": "📎", "office365": "📎", "microsoft": "📎", "google": "🔎",
        "googleworkspace": "📎",
        "linkedin": "💼", "grammarly": "✍️", "duolingo": "🦉", "masterclass": "🎓",
        "coursera": "🎓", "skillshare": "🎓",
    ]

    /// Brand keys, longest first, so "applemusic" wins over "apple".
    static let brandsByLength: [String] = brandEmoji.keys.sorted { $0.count > $1.count }

    /// Normalize a name for matching: lowercase, strip non-alphanumerics.
    public static func normalize(_ name: String) -> String {
        name.lowercased().unicodeScalars.filter {
            CharacterSet.alphanumerics.contains($0)
        }.map(String.init).joined()
    }

    /// The per-brand emoji for a name, or nil when no brand matched.
    public static func brand(_ name: String) -> String? {
        let key = normalize(name)
        if let hit = brandEmoji[key] { return hit }
        for b in brandsByLength where key.contains(b) { return brandEmoji[b] }
        return nil
    }

    /// Emoji for a subscription/merchant (per-brand → category → generic).
    public static func emoji(_ name: String, category: String? = nil) -> String {
        if let hit = brand(name) { return hit }
        return category == "Subscriptions" ? "📱" : "🔁"
    }
}

File: android/core/src/main/kotlin/app/fihaven/core/logic/SubscriptionIcons.kt
package app.fihaven.core.logic

/**
 * Map a subscription / merchant name to a recognizable emoji. The web app also
 * renders real SVG brand logos, but native mirrors only the emoji layer
 * (per-brand → category → generic). Keep the brand table in sync with the web
 * `subscriptionIcons.js` and iOS `SubscriptionIcons.swift`.
 */
object SubscriptionIcons {
    /** Recognizable per-brand emoji, keyed by the normalized name. */
    val BRAND_EMOJI: Map<String, String> = mapOf(
        "netflix" to "🎬", "disney" to "🏰", "disneyplus" to "🏰", "hulu" to "📺", "max" to "📺",
        "hbomax" to "📺", "hbo" to "📺", "peacock" to "🦚", "paramount" to "⛰️", "paramountplus" to "⛰️",
        "appletv" to "🍏", "appletvplus" to "🍏", "apple" to "🍏",
        "icloud" to "☁️", "icloudstorage" to "☁️", "icloudplus" to "☁️", "googleone" to "☁️",
        "dropbox" to "📦", "onedrive" to "☁️",
        "spotify" to "🎵", "applemusic" to "🎵", "youtubemusic" to "🎵", "pandora" to "🎵",
        "tidal" to "🎵", "music" to "🎵", "deezer" to "🎵",
        "youtube" to "▶️", "youtubepremium" to "▶️", "youtubetv" to "📺", "twitch" to "🎮",
        "xboxgamepass" to "🎮", "playstationplus" to "🎮", "playstation" to "🎮", "nintendo" to "🎮",
        "nintendoswitchonline" to "🎮",
        "amazon" to "📦", "amazonprime" to "📦", "primevideo" to "📦",
        "audible" to "🎧", "kindle" to "📚", "kindleunlimited" to "📚",
        "nyt" to "📰", "nytimes" to "📰", "wsj" to "📰", "washingtonpost" to "📰", "medium" to "📰",
        "theathletic" to "📰",
        "patreon" to "🅿️", "substack" to "📩", "notion" to "📝", "evernote" to "📝", "obsidian" to "📝",
        "adobe" to "🎨", "adobecc" to "🎨", "canva" to "🎨", "figma" to "🎨", "lightroom" to "🎨",
        "onepassword" to "🔐", "bitwarden" to "🔐", "lastpass" to "🔐", "dashlane" to "🔐",
        "nordvpn" to "🛡️", "expressvpn" to "🛡️", "protonvpn" to "🛡️", "proton" to "🛡️",
        "chatgpt" to "🤖", "openai" to "🤖", "claude" to "🤖", "anthropic" to "🤖", "midjourney" to "🤖",
        "perplexity" to "🤖",
        "github" to "🐙", "githubcopilot" to "🐙", "githubpro" to "🐙",
        "peloton" to "🚲", "strava" to "🏃", "calm" to "🧘", "headspace" to "🧘", "fitbit" to "⌚",
        "whoop" to "⌚",
        "crunchyroll" to "🍥", "openbubbles" to "💬", "slack" to "💬", "discord" to "💬", "zoom" to "🎥",
        "microsoft365" to "📎", "office365" to "📎", "microsoft" to "📎", "google" to "🔎",
        "googleworkspace" to "📎",
        "linkedin" to "💼", "grammarly" to "✍️", "duolingo" to "🦉", "masterclass" to "🎓",
        "coursera" to "🎓", "skillshare" to "🎓",
    )

    /** Brand keys, longest first, so "applemusic" wins over "apple". */
    private val brandsByLength: List<String> = BRAND_EMOJI.keys.sortedByDescending { it.length }

    /** Normalize a name for matching: lowercase, strip non-alphanumerics. */
    fun normalize(name: String?): String =
        (name ?: "").lowercase().filter { it in 'a'..'z' || it in '0'..'9' }

    /** The per-brand emoji for a name, or null when no brand matched. */
    fun brand(name: String?): String? {
        val key = normalize(name)
        BRAND_EMOJI[key]?.let { return it }
        for (b in brandsByLength) if (key.contains(b)) return BRAND_EMOJI[b]
        return null
    }

    /** Emoji for a subscription/merchant (per-brand → generic recurring). */
    fun emoji(name: String?, category: String? = null): String =
        brand(name) ?: "🔁"
}
